internal/task/error.go

```diff
@@ -0,0 +1,93 @@
+package task
+
+import (
+	"errors"
+	"fmt"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ErrorCode represents the type of error that occurred
+type ErrorCode int
+
+const (
+	// ErrInvalidArgument indicates that the provided arguments are invalid
+	ErrInvalidArgument ErrorCode = iota
+	// ErrNotFound indicates that the requested resource was not found
+	ErrNotFound
+	// ErrAlreadyTerminated indicates that the task has already terminated
+	ErrAlreadyTerminated
+	// ErrFailedPrecondition indicates that the task is in a failed precondition state
+	ErrFailedPrecondition
+	// ErrInternal indicates an internal system error
+	ErrInternal
+)
+
+// TaskError represents an error that occurred during task management
+type TaskError struct {
+	Code    ErrorCode
+	Message string
+	Err     error
+}
+
+func (e *TaskError) Error() string {
+	if e.Err != nil {
+		return e.Message + ": " + e.Err.Error()
+	}
+	return e.Message
+}
+
+func (e *TaskError) Unwrap() error {
+	return e.Err
+}
+
+// NewTaskError creates a new TaskError with optional formatting
+func NewTaskError(code ErrorCode, format string, args ...any) *TaskError {
+	return &TaskError{
+		Code:    code,
+		Message: fmt.Sprintf(format, args...),
+	}
+}
+
+// NewTaskErrorWithErr creates a new TaskError with an underlying error
+func NewTaskErrorWithErr(code ErrorCode, format string, err error, args ...any) *TaskError {
+	return &TaskError{
+		Code:    code,
+		Message: fmt.Sprintf(format, args...),
+		Err:     err,
+	}
+}
+
+// IsTaskError checks if an error is a TaskError
+func IsTaskError(err error) (*TaskError, bool) {
+	var taskErr *TaskError
+	if errors.As(err, &taskErr) {
+		return taskErr, true
+	}
+	return nil, false
+}
+
+// TaskErrorToGRPC converts a TaskError to a gRPC status error
+func TaskErrorToGRPC(err error) error {
+	if taskErr, ok := IsTaskError(err); ok {
```

Comment:
can't you use errors.As api instead of calling `IsTaskError` that instead of returning a bool also returns the error?

Reply:
Yes that is a much cleaner approach